src/main/Main.hs

```diff
@@ -865,7 +880,10 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
                     -- Unlock before transferring control away, whether using docker or not:
                     (Just $ munlockFile lk)
                     (runStackTGlobal manager (lcConfig lc) go $
-                        exec plainEnvSettings cmd args)
+                        Nix.execWithOptionalShell
```

Comment:
You might be able to use `reexecWithOptionalShell` here.  The Docker case has switched to always using `reexecWithOptionalContainer` (even with `stack exec --plain`) and you can probably do the same.  After switching, that may open the opportunity to simplify `Nix.rexecWithOptionalShell` and `Nix.execWithOptionalShell` a bit.